validator/client/runner.go

```diff
@@ -24,123 +24,171 @@ import (
+// a) fetch and perform validator roles every slot,
+// b) periodically check if the beacon node is healthy and perform a switchover if possible
+// c) push proposer settings when appropriate
+// d) process events
+// e) act on account changes
 func run(ctx context.Context, v iface.Validator) {
 	cleanup := v.Done
 	defer cleanup()
 
-	headSlot, err := initializeValidatorAndGetHeadSlot(ctx, v)
-	if err != nil {
-		return // Exit if context is canceled.
+	if err := initializeValidator(ctx, v); err != nil {
+		log.WithError(err).Error("Could not initialize validator")
+		return
 	}
-	if err := v.UpdateDuties(ctx, headSlot); err != nil {
-		handleAssignmentError(err, headSlot)
+	headSlot, err := v.CanonicalHeadSlot(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get canonical head slot")
+		return
 	}
-	eventsChan := make(chan *event.Event, 1)
-	healthTracker := v.HealthTracker()
-	runHealthCheckRoutine(ctx, v, eventsChan)
 
-	accountsChangedChan := make(chan [][fieldparams.BLSPubkeyLength]byte, 1)
 	km, err := v.Keymanager()
 	if err != nil {
-		log.WithError(err).Fatal("Could not get keymanager")
+		log.WithError(err).Error("Could not get keymanager")
+		return
 	}
+	accountsChangedChan := make(chan [][fieldparams.BLSPubkeyLength]byte, 1)
 	sub := km.SubscribeAccountChanges(accountsChangedChan)
-	// check if proposer settings is still nil
-	// Set properties on the beacon node like the fee recipient for validators that are being used & active.
+
 	if v.ProposerSettings() == nil {
 		log.Warn("Validator client started without proposer settings such as fee recipient" +
 			" and will continue to use settings provided in the beacon node.")
 	}
-	deadline := time.Now().Add(5 * time.Minute)
-	if err := v.PushProposerSettings(ctx, km, headSlot, deadline); err != nil {
-		if errors.Is(err, ErrBuilderValidatorRegistration) {
-			log.WithError(err).Warn("Push proposer settings error")
-		} else {
-			log.WithError(err).Fatal("Failed to update proposer settings") // allow fatal. skipcq
+
+	pushProposerSettingsChan := make(chan primitives.Slot, 1)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case slot := <-pushProposerSettingsChan:
+				go func() {
+					psCtx, psCancel := context.WithDeadline(ctx, time.Now().Add(5*time.Minute))
+					if err := v.PushProposerSettings(psCtx, km, slot); err != nil {
+						log.WithError(err).Warn("Failed to update proposer settings")
+					}
+					// release resources
+					psCancel()
+				}()
+			}
 		}
+	}()
+	pushProposerSettingsChan <- headSlot
+
+	updateDutiesNeeded := false
+	if err = v.UpdateDuties(ctx, headSlot); err != nil {
+		handleUpdateDutiesError(err, headSlot)
+		updateDutiesNeeded = true
 	}
+
+	eventsChan := make(chan *event.Event, 1)
+	go v.StartEventStream(ctx, event.DefaultEventTopics, eventsChan)
+
+	var (
+		slotSpan             *trace.Span
+		slotCtx              context.Context
+		slotCancel           context.CancelFunc
+		initializationNeeded = false
```

Comment:
maybe we should just have some kind of validator client status, feels hacky to use a variable like this

Reply:
I agree it might be a good idea, but would probably result in a much different design, further complicating the already non-trivial PR.